### ISSUE TITLE
4.6.0: ssh configuration in OpenShift 3.11

### ISSUE THREAD

**Issue by jkatz** [CrunchyData/postgres-operator#2289]
We are having issues with the backup piece of the operator. This time it is the `stanza-create` job failing. We think we have narrowed it down to a bug in the ssh config between  `backrest-shared-repo` and the `postgres` pod.  (maybe also a permissions issue like above?)

When we ssh from the `backrest-shared-repo` pod: 
``` bash
bash-4.4$ ssh -o LogLevel=debug -o Compression=no -o PasswordAuthentication=no pgbackrest@10.150.4.66
debug1: Reading configuration data //.ssh/config
debug1: //.ssh/config line 1: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: Reading configuration data /etc/ssh/ssh_config.d/05-redhat.conf
debug1: Reading configuration data /etc/crypto-policies/back-ends/openssh.config
debug1: configuration requests final Match pass
debug1: re-parsing configuration
debug1: Reading configuration data //.ssh/config
debug1: //.ssh/config line 1: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: Reading configuration data /etc/ssh/ssh_config.d/05-redhat.conf
debug1: Reading configuration data /etc/crypto-policies/back-ends/openssh.config
debug1: FIPS mode initialized
debug1: Connecting to 10.150.4.66 [10.150.4.66] port 2022.
debug1: Connection established.
debug1: identity file /tmp/id_ed25519 type -1
debug1: identity file /tmp/id_ed25519-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.0
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.0
debug1: match: OpenSSH_8.0 pat OpenSSH* compat 0x04000000
debug1: Authenticating to10.150.4.66:2022 as 'pgbackrest'
debug1: SSH2_MSG_KEXINIT sent
Connection reset by 10.150.4.66 port 2022
```

I modified the configs on the ssh server in the postgres container to get the debug logs: 
```bash
debug1: FIPS mode initialized
debug1: inetd sockets after dupping: 4, 4
Connection from 10.140.5.30 port 41354 on 10.150.4.66 port 2022
debug1: Local version string SSH-2.0-OpenSSH_8.0
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.0
debug1: match: OpenSSH_8.0 pat OpenSSH* compat 0x04000000
debug1: sshd_selinux_change_privsep_preauth_context: Failed to open SELinux context file [preauth]
debug1: list_hostkey_types:  [preauth]
No supported key exchange algorithms [preauth]
debug1: do_cleanup [preauth]
debug1: monitor_read_log: child log fd closed
debug1: do_cleanup
debug1: Killing privsep child 49863
```

While we were exploring this we also saw this error in the ssh server logs. Dont know if it is a warning or actually an issue for the SSHD server: 
```Couldn't create pid file "/var/run/sshd.pid": Permission denied```

We have verified that ssh is working in the 4.5.1 operator we are still running in the cluster. Does someone know anything that changed in the new version that might have broken SSH? Our theory are either incompatible config with FIPS or bad permissions on the keys in the container.

_Originally posted by @Vermillionaire in https://github.com/CrunchyData/postgres-operator/issues/2284#issuecomment-780968058_

**Comment by jkatz** [comment]
@Vermillionaire I need some more information:

- Are you running in FIPS mode for 4.5?
- What kinds of keys are you using? Are you using ED25519 (Operator default 4.2+), ECDSA, or RSA?
- Did you customize the `sshd_config` at all?
- Which base image are you using?

This also feels like it's missing a few logs that can identify what the issue actually is. Some things that would help:

- The logs of the pgBackRest failure messages
- Determine if sshd is running on the pgBackRest repository Pod
- Did you add any other restrictions or security context rules on the Pods?

**Comment by Vermillionaire** [comment]
* Yes, we are running FIPS for 4.5. We have 4.5 and 4.6 in the same cluster, separate namespaces. 
* We have have not touched the keys that are being used (ED25519)
* We modified the sshd_conf to get some logs out and change the pid location: 
``` 
#	$OpenBSD: sshd_config,v 1.100 2016/08/15 12:32:04 naddy Exp $

# This is the sshd server system-wide configuration file.  See
# sshd_config(5) for more information.

# This sshd was compiled with PATH=/usr/local/bin:/usr/bin

# The strategy used for options in the default sshd_config shipped with
# OpenSSH is to specify options with their default value where
# possible, but leave them commented.  Uncommented options override the
# default value.

# If you want to change the port on a SELinux system, you have to tell
# SELinux about this change.
# semanage port -a -t ssh_port_t -p tcp #PORTNUMBER
#
Port 2022
#AddressFamily any
#ListenAddress 0.0.0.0
#ListenAddress ::

HostKey /sshd/ssh_host_ed25519_key

# Ciphers and keying
#RekeyLimit default none

# Logging
#SyslogFacility AUTH
SyslogFacility AUTHPRIV
LogLevel DEBUG

# Authentication:

#LoginGraceTime 2m
PermitRootLogin no
StrictModes no
#MaxAuthTries 6
#MaxSessions 10

PubkeyAuthentication yes

# The default is to check both .ssh/authorized_keys and .ssh/authorized_keys2
# but this is overridden so installations will only check .ssh/authorized_keys
#AuthorizedKeysFile	/pgconf/authorized_keys
AuthorizedKeysFile	/sshd/authorized_keys

#AuthorizedPrincipalsFile none

#AuthorizedKeysCommand none
#AuthorizedKeysCommandUser nobody

# For this to work you will also need host keys in /etc/ssh/ssh_known_hosts
#HostbasedAuthentication no
# Change to yes if you don't trust ~/.ssh/known_hosts for
# HostbasedAuthentication
#IgnoreUserKnownHosts no
# Don't read the user's ~/.rhosts and ~/.shosts files
#IgnoreRhosts yes

# To disable tunneled clear text passwords, change to no here!
#PasswordAuthentication yes
#PermitEmptyPasswords no
PasswordAuthentication no

# Change to no to disable s/key passwords
ChallengeResponseAuthentication yes
#ChallengeResponseAuthentication no

# Kerberos options
#KerberosAuthentication no
#KerberosOrLocalPasswd yes
#KerberosTicketCleanup yes
#KerberosGetAFSToken no
#KerberosUseKuserok yes

# GSSAPI options
#GSSAPIAuthentication yes
#GSSAPICleanupCredentials no
#GSSAPIStrictAcceptorCheck yes
#GSSAPIKeyExchange no
#GSSAPIEnablek5users no

# Set this to 'yes' to enable PAM authentication, account processing,
# and session processing. If this is enabled, PAM authentication will
# be allowed through the ChallengeResponseAuthentication and
# PasswordAuthentication.  Depending on your PAM configuration,
# PAM authentication via ChallengeResponseAuthentication may bypass
# the setting of "PermitRootLogin without-password".
# If you just want the PAM account and session checks to run without
# PAM authentication, then enable this but set PasswordAuthentication
# and ChallengeResponseAuthentication to 'no'.
# WARNING: 'UsePAM no' is not supported in Red Hat Enterprise Linux and may cause several
# problems.
UsePAM yes

#AllowAgentForwarding yes
#AllowTcpForwarding yes
#GatewayPorts no
X11Forwarding yes
#X11DisplayOffset 10
#X11UseLocalhost yes
#PermitTTY yes
#PrintMotd yes
#PrintLastLog yes
#TCPKeepAlive yes
#UseLogin no
#PermitUserEnvironment no
#Compression delayed
#ClientAliveInterval 0
#ClientAliveCountMax 3
#ShowPatchLevel no
#UseDNS yes
PidFile /tmp/sshd.pid
#MaxStartups 10:30:100
#PermitTunnel no
#ChrootDirectory none
#VersionAddendum none

# no default banner path
#Banner none

# Accept locale-related environment variables
AcceptEnv LANG LC_CTYPE LC_NUMERIC LC_TIME LC_COLLATE LC_MONETARY LC_MESSAGES
AcceptEnv LC_PAPER LC_NAME LC_ADDRESS LC_TELEPHONE LC_MEASUREMENT
AcceptEnv LC_IDENTIFICATION LC_ALL LANGUAGE
AcceptEnv XMODIFIERS

# override default of no subsystems
Subsystem	sftp	/usr/libexec/openssh/sftp-server

# Example of overriding settings on a per-user basis
#Match User anoncvs
#	X11Forwarding no
#	AllowTcpForwarding no
#	PermitTTY no
#	ForceCommand cvs server
```

This is the log output from the stanza job. There was no failure messages in the backrest shared repo itself: 
```
Wed Feb 17 22:32:53 UTC 2021 INFO: Image mode found: pgbackrest
Wed Feb 17 22:32:53 UTC 2021 INFO: Starting in 'pgbackrest' mode
time="2021-02-17T22:32:54Z" level=info msg="crunchy-pgbackrest starts"
time="2021-02-17T22:32:54Z" level=info msg="debug flag set to %tfalse"
time="2021-02-17T22:32:54Z" level=info msg="backrest stanza-create command requested"
time="2021-02-17T22:32:54Z" level=info msg="command to execute is [pgbackrest stanza-create  --db-host=10.150.4.66 --db-path=/pgdata/postgres]"
time="2021-02-17T22:32:54Z" level=info msg="output=[]"
time="2021-02-17T22:32:54Z" level=info msg="stderr=[WARN: unable to check pg-1: [UnknownError] remote-0 process on '10.128.5.61' terminated unexpectedly [255]\nERROR: [056]: unable to find primary cluster - cannot proceed\n]"
time="2021-02-17T22:32:54Z" level=fatal msg="command terminated with exit code 56"
```

* This is the image tag for postgres: `crunchydata-postgres-ha-centos8-12.5-4.6.0-0.1`
* This is the security context on the postgres pod: `openshift.io/scc: restricted`

The sshd process was running when we checked the pod initially. We stopped it and tweaked the config a bit to get log outputs that show the errors in the original post.

**Comment by mjrussell** [comment]
@jkatz @andrewlecuyer let us know if theres any other info we can provide. We are really hoping to figure this out so we can get backups and replications working properly in our setup (replicas dont come online until create stanza succeeds)

**Comment by jkatz** [comment]
We still need to investigate if there is an actual bug, if there is an environmental issue, etc. If you are looking for more immediate support, please review the various [support](https://access.crunchydata.com/documentation/postgres-operator/latest/support/) options.

I would also suggest trying on 4.6.1 to see if any of the bug fixes or changes have resolved the issue.

**Comment by andrewlecuyer** [comment]
@mjrussell can you provide any details regarding how you're enabling FIPS?

**Comment by mjrussell** [comment]
Hey all, sorry for the delay on our side.


We were able to figure this out and get it working. FIPS was absolutely the problem, as apparently edsa is not supported on rhel8 https://access.redhat.com/solutions/3643252. I'm not sure why we didn't have this problem on rhel7, but I was able to follow the section about changing the ssh key to rsa here https://access.crunchydata.com/documentation/postgres-operator/4.5.0/custom-resources/.

I saw this was removed from the documentation, I probably would have had a much harder time without it. Maybe it should be added back but in a troubleshooting guide about FIPS?
